Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,265 @@
+                        metadata: [
+                            self.configuration.logging.signalKey: "\(unixSignal)",
+                        ]
+                    )
+
+                    // We have to shutdown the services in reverse. To do this
+                    // we are going to signal each child task the graceful shutdown and then wait for
+                    // its exit.
+                    for (gracefulShutdownIndex, gracefulShutdownManager) in gracefulShutdownManagers.lazy.enumerated().reversed() {
+                        self.logger.debug(
+                            "Triggering graceful shutdown for service",
+                            metadata: [
+                                self.configuration.logging.serviceKey: "\(self.services[gracefulShutdownIndex])",
+                            ]
+                        )
+
+                        await gracefulShutdownManager.shutdownGracefully()
+
+                        let result = await group.next()
+
+                        switch result {
+                        case .serviceFinished(let service, let index):
+                            if index == gracefulShutdownIndex {
+                                // The service that we signalled graceful shutdown did exit/
+                                // We can continue to the next one.
+                                self.logger.debug(
+                                    "Service finished",
+                                    metadata: [
+                                        self.configuration.logging.serviceKey: "\(service)",
+                                    ]
+                                )
+                                continue
+                            } else {
+                                // Another service exited unexpectedly
+                                self.logger.error(
+                                    "Service finished unexpectedly during graceful shutdown. Cancelling all other services now",
+                                    metadata: [
+                                        self.configuration.logging.serviceKey: "\(service)",
+                                    ]
+                                )
+
+                                group.cancelAll()
+                                return .failure(ServiceRunnerError.serviceFinishedUnexpectedly())
+                            }
+
+                        case .serviceThrew(let service, _, let error):
+                            self.logger.error(
+                                "Service threw error during graceful shutdown. Cancelling all other services now",
+                                metadata: [
+                                    self.configuration.logging.serviceKey: "\(service)",
+                                    self.configuration.logging.errorKey: "\(error)",
+                                ]
+                            )
+                            group.cancelAll()
```

Comment:
if one service throws on shutdown we want to escalate all following graceful shutdowns to cancel. this seems quite aggressive